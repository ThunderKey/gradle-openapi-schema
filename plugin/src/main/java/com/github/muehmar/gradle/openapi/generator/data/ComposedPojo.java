package com.github.muehmar.gradle.openapi.generator.data;

import ch.bluecare.commons.data.PList;

import java.util.Objects;

/**
 * Represents a composition of other definitions, i.e. pojos. Distinguishes between different types
 * of compositions, see {@link CompositionType}.
 */
public class ComposedPojo {
  private final CompositionType type;
  private final PList<String> pojoNames;
  private final PList<PojoMember> members;

  public enum CompositionType {
    ALL_OF,
    ANY_OF,
    ONE_OF;
  }

  public ComposedPojo(CompositionType type, PList<String> pojoNames, PList<PojoMember> members) {
    this.type = type;
    this.pojoNames = pojoNames;
    this.members = members;
  }

  public CompositionType getType() {
    return type;
  }

  public PList<String> getPojoNames() {
    return pojoNames;
  }

  public PList<PojoMember> getMembers() {
    return members;
  }

  @Override
  public boolean equals(Object o) {

    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ComposedPojo that = (ComposedPojo) o;
    return type == that.type
        && Objects.equals(pojoNames, that.pojoNames)
        && Objects.equals(members, that.members);
  }

  @Override
  public int hashCode() {
    return Objects.hash(type, pojoNames, members);
  }

  @Override
  public String toString() {
    return "ComposedPojo{"
        + "type="
        + type
        + ", pojoNames="
        + pojoNames
        + ", members="
        + members
        + '}';
  }
}
