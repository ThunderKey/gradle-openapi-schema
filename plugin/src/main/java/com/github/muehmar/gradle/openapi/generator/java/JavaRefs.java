package com.github.muehmar.gradle.openapi.generator.java;

public class JavaRefs {
  private JavaRefs() {}

  public static final String LOCAL_DATE = "java.time.LocalDate";
  public static final String OPTIONAL = "java.util.Optional";
}
