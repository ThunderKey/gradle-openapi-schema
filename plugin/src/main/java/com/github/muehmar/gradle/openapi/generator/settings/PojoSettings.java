package com.github.muehmar.gradle.openapi.generator.settings;

import ch.bluecare.commons.data.PList;
import io.github.muehmar.pojoextension.annotations.PojoExtension;
import java.io.Serializable;
import java.util.List;
import java.util.Objects;

@PojoExtension
public class PojoSettings extends PojoSettingsExtension implements Serializable {
  private final JsonSupport jsonSupport;
  private final String packageName;
  private final String suffix;
  private final boolean enableSafeBuilder;
  private final boolean enableConstraints;
  private final List<ClassTypeMapping> classTypeMappings;
  private final List<FormatTypeMapping> formatTypeMappings;
  private final EnumDescriptionSettings enumDescriptionSettings;

  PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      List<ClassTypeMapping> classTypeMappings,
      List<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings;
    this.formatTypeMappings = formatTypeMappings;
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public PojoSettings(
      JsonSupport jsonSupport,
      String packageName,
      String suffix,
      boolean enableSafeBuilder,
      boolean enableConstraints,
      PList<ClassTypeMapping> classTypeMappings,
      PList<FormatTypeMapping> formatTypeMappings,
      EnumDescriptionSettings enumDescriptionSettings) {
    this.jsonSupport = jsonSupport;
    this.packageName = packageName;
    this.suffix = suffix;
    this.enableSafeBuilder = enableSafeBuilder;
    this.enableConstraints = enableConstraints;
    this.classTypeMappings = classTypeMappings.toArrayList();
    this.formatTypeMappings = formatTypeMappings.toArrayList();
    this.enumDescriptionSettings = enumDescriptionSettings;
  }

  public String getPackageName() {
    return packageName;
  }

  public String getSuffix() {
    return suffix;
  }

  public JsonSupport getJsonSupport() {
    return jsonSupport;
  }

  public boolean isJacksonJson() {
    return jsonSupport.equals(JsonSupport.JACKSON);
  }

  public boolean isEnableConstraints() {
    return enableConstraints;
  }

  public PList<ClassTypeMapping> getClassTypeMappings() {
    return PList.fromIter(classTypeMappings);
  }

  public PList<FormatTypeMapping> getFormatTypeMappings() {
    return PList.fromIter(formatTypeMappings);
  }

  public boolean isEnableSafeBuilder() {
    return enableSafeBuilder;
  }

  public boolean isDisableSafeBuilder() {
    return !isEnableSafeBuilder();
  }

  public EnumDescriptionSettings getEnumDescriptionSettings() {
    return enumDescriptionSettings;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    PojoSettings that = (PojoSettings) o;
    return enableSafeBuilder == that.enableSafeBuilder
        && enableConstraints == that.enableConstraints
        && jsonSupport == that.jsonSupport
        && Objects.equals(packageName, that.packageName)
        && Objects.equals(suffix, that.suffix)
        && Objects.equals(classTypeMappings, that.classTypeMappings)
        && Objects.equals(formatTypeMappings, that.formatTypeMappings)
        && Objects.equals(enumDescriptionSettings, that.enumDescriptionSettings);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        jsonSupport,
        packageName,
        suffix,
        enableSafeBuilder,
        enableConstraints,
        classTypeMappings,
        formatTypeMappings,
        enumDescriptionSettings);
  }

  @Override
  public String toString() {
    return "PojoSettings{"
        + "jsonSupport="
        + jsonSupport
        + ", packageName='"
        + packageName
        + '\''
        + ", suffix='"
        + suffix
        + '\''
        + ", enableSafeBuilder="
        + enableSafeBuilder
        + ", enableConstraints="
        + enableConstraints
        + ", classTypeMappings="
        + classTypeMappings
        + ", formatTypeMappings="
        + formatTypeMappings
        + ", enumDescriptionExtraction="
        + enumDescriptionSettings
        + '}';
  }
}
