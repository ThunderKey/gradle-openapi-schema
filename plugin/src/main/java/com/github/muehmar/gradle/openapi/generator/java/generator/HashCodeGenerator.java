package com.github.muehmar.gradle.openapi.generator.java.generator;

import static com.github.muehmar.gradle.openapi.util.Functions.allExceptFirst;
import static io.github.muehmar.pojoextension.generator.impl.JavaModifier.PUBLIC;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.data.Name;
import com.github.muehmar.gradle.openapi.generator.data.Pojo;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.JavaResolver;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.pojoextension.generator.Generator;
import io.github.muehmar.pojoextension.generator.impl.gen.MethodGen;
import io.github.muehmar.pojoextension.generator.impl.gen.MethodGenBuilder;
import io.github.muehmar.pojoextension.generator.writer.Writer;

public class HashCodeGenerator {
  private static final JavaResolver RESOLVER = new JavaResolver();

  private HashCodeGenerator() {}

  public static Generator<Pojo, PojoSettings> hashCodeMethod() {
    final MethodGen<Pojo, PojoSettings> method =
        MethodGenBuilder.<Pojo, PojoSettings>create()
            .modifiers(PUBLIC)
            .noGenericTypes()
            .returnType("int")
            .methodName("hashCode")
            .noArguments()
            .content(hashCodeMethodContent())
            .build();
    return AnnotationGenerator.<Pojo, PojoSettings>override().append(method);
  }

  private static Generator<Pojo, PojoSettings> hashCodeMethodContent() {
    return (pojo, s, w) -> {
      final PList<String> fieldNames =
          pojo.getMembers()
              .flatMap(
                  member -> {
                    final Name memberName = member.memberName(RESOLVER);
                    if (member.isRequiredAndNullable()) {
                      final String requiredNullableFlagName =
                          String.format("is%sPresent", memberName.startUpperCase());
                      return PList.of(memberName.asString(), requiredNullableFlagName);
                    } else if (member.isOptionalAndNullable()) {
                      final String optionalNullableFlagName =
                          String.format("is%sNull", memberName.startUpperCase());
                      return PList.of(memberName.asString(), optionalNullableFlagName);
                    } else {
                      return PList.single(memberName.asString());
                    }
                  });

      final Writer writerStartPrinted = w.println("return Objects.hash(");

      final PList<String> mappedFieldNames =
          fieldNames.reverse().zipWithIndex().map(allExceptFirst(name -> name + ",")).reverse();

      return mappedFieldNames
          .foldLeft(writerStartPrinted, (writer, name) -> writer.tab(1).println(name))
          .println(");")
          .ref(JavaRefs.JAVA_UTIL_OBJECTS);
    };
  }
}
