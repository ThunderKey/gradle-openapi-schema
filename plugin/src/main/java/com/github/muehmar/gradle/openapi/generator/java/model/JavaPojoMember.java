package com.github.muehmar.gradle.openapi.generator.java.model;

import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.Name;
import com.github.muehmar.gradle.openapi.generator.model.Necessity;
import com.github.muehmar.gradle.openapi.generator.model.Nullability;
import com.github.muehmar.gradle.openapi.generator.model.PojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.GetterSuffixes;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import lombok.EqualsAndHashCode;
import lombok.ToString;

@EqualsAndHashCode
@ToString
public class JavaPojoMember {
  private final Name name;
  private final String description;
  private final JavaType javaType;
  private final Necessity necessity;
  private final Nullability nullability;

  private JavaPojoMember(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    this.javaType = javaType;
    this.name = name;
    this.description = description;
    this.necessity = necessity;
    this.nullability = nullability;
  }

  public static JavaPojoMember of(
      Name name,
      String description,
      JavaType javaType,
      Necessity necessity,
      Nullability nullability) {
    return new JavaPojoMember(name, description, javaType, necessity, nullability);
  }

  public static JavaPojoMember wrap(PojoMember pojoMember, TypeMappings typeMappings) {
    final JavaType javaType = JavaType.wrap(pojoMember.getType(), typeMappings);
    return new JavaPojoMember(
        pojoMember.getName(),
        pojoMember.getDescription(),
        javaType,
        pojoMember.getNecessity(),
        pojoMember.getNullability());
  }

  public Nullability getNullability() {
    return nullability;
  }

  public Necessity getNecessity() {
    return necessity;
  }

  public String getDescription() {
    return description;
  }

  public Name getName() {
    return name;
  }

  public JavaType getJavaType() {
    return javaType;
  }

  public boolean isOptional() {
    return necessity.isOptional();
  }

  public boolean isRequired() {
    return necessity.isRequired();
  }

  public boolean isNullable() {
    return nullability.isNullable();
  }

  public boolean isNotNullable() {
    return nullability.isNotNullable();
  }

  public boolean isRequiredAndNullable() {
    return isRequired() && isNullable();
  }

  public boolean isRequiredAndNotNullable() {
    return isRequired() && isNotNullable();
  }

  public boolean isOptionalAndNullable() {
    return isOptional() && isNullable();
  }

  public boolean isOptionalAndNotNullable() {
    return isOptional() && isNotNullable();
  }

  public Name getGetterName() {
    return prefixedMethodName("get");
  }

  public Name getGetterNameWithSuffix(PojoSettings settings) {
    return getGetterName().append(determineSuffix(settings));
  }

  private String determineSuffix(PojoSettings settings) {
    final GetterSuffixes getterSuffixes = settings.getGetterSuffixes();
    if (isRequiredAndNotNullable()) {
      return getterSuffixes.getRequiredSuffix();
    } else if (isRequiredAndNullable()) {
      return getterSuffixes.getRequiredNullableSuffix();
    } else if (isOptionalAndNotNullable()) {
      return getterSuffixes.getOptionalSuffix();
    } else {
      return getterSuffixes.getOptionalNullableSuffix();
    }
  }

  public Name getWitherName() {
    return prefixedMethodName("with");
  }

  public Name prefixedMethodName(String prefix) {
    if (prefix.isEmpty()) {
      return name;
    } else {
      return name.startUpperCase().prefix(prefix);
    }
  }
}
