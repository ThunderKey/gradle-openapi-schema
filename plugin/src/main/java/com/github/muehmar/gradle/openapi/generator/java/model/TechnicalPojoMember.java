package com.github.muehmar.gradle.openapi.generator.java.model;

import static com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties.additionalPropertiesName;

import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaBooleanType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaMapType;
import com.github.muehmar.gradle.openapi.generator.java.model.type.JavaType;
import com.github.muehmar.gradle.openapi.generator.model.type.MapType;
import com.github.muehmar.gradle.openapi.generator.model.type.StringType;
import com.github.muehmar.gradle.openapi.generator.settings.TypeMappings;
import lombok.Value;

/**
 * Represents a technical member of a pojo, i.e. a generated pojo contains this member as class
 * member. It may also be a helper flag to represent an additional state of the actual member
 * represented by {@link JavaPojoMember}. This means, an actual member {@link JavaPojoMember} may
 * result in more than one {@link TechnicalPojoMember} in a pojo to be able to fully represent the
 * state of a member (for example tristate members aka optional nullable members).
 */
@Value
public class TechnicalPojoMember {
  JavaIdentifier name;
  JavaType javaType;

  public static TechnicalPojoMember additionalProperties(JavaType valueType) {
    final MapType mapType = MapType.ofKeyAndValueType(StringType.noFormat(), valueType.getType());
    final JavaMapType javaMapType = JavaMapType.wrap(mapType, TypeMappings.empty());
    return new TechnicalPojoMember(additionalPropertiesName(), javaMapType);
  }

  public static TechnicalPojoMember isPresentFlagMember(JavaIdentifier name) {
    return new TechnicalPojoMember(name, JavaBooleanType.create());
  }

  public static TechnicalPojoMember isNullFlagMember(JavaIdentifier name) {
    return new TechnicalPojoMember(name, JavaBooleanType.create());
  }

  public static TechnicalPojoMember wrapPojoMember(JavaPojoMember member) {
    return new TechnicalPojoMember(member.getName().asIdentifier(), member.getJavaType());
  }
}
