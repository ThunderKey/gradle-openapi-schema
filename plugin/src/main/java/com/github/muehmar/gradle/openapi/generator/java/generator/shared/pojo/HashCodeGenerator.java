package com.github.muehmar.gradle.openapi.generator.java.generator.shared.pojo;

import static com.github.muehmar.gradle.openapi.util.Functions.allExceptFirst;
import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.generator.pojo.AnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaAdditionalProperties;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaIdentifier;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import java.util.Optional;
import lombok.Value;

public class HashCodeGenerator {
  private HashCodeGenerator() {}

  public static Generator<HashCodeContent, PojoSettings> hashCodeMethod() {
    final Generator<HashCodeContent, PojoSettings> method =
        JavaGenerators.<HashCodeContent, PojoSettings>methodGen()
            .modifiers(PUBLIC)
            .noGenericTypes()
            .returnType("int")
            .methodName("hashCode")
            .noArguments()
            .content(hashCodeMethodContent())
            .build();
    return AnnotationGenerator.<HashCodeContent, PojoSettings>override()
        .append(method)
        .append(arraysRefGenerator());
  }

  private static Generator<HashCodeContent, PojoSettings> hashCodeMethodContent() {
    return (content, s, w) -> {
      final Writer writerStartPrinted = w.println("return Objects.hash(");

      final PList<String> additionalPropertiesArgument =
          PList.fromOptional(
                  content
                      .getAdditionalProperties()
                      .map(ignore -> JavaAdditionalProperties.getPropertyName()))
              .map(JavaIdentifier::asString);

      return content
          .getMembers()
          .map(HashCodeMember::new)
          .flatMap(HashCodeMember::toHashCodeMethodArguments)
          .concat(additionalPropertiesArgument)
          .reverse()
          .zipWithIndex()
          .map(allExceptFirst(arg -> arg.concat(",")))
          .reverse()
          .foldLeft(writerStartPrinted, (writer, name) -> writer.tab(1).println(name))
          .println(");")
          .ref(JavaRefs.JAVA_UTIL_OBJECTS);
    };
  }

  private static Generator<HashCodeContent, PojoSettings> arraysRefGenerator() {
    return Generator.<HashCodeContent, PojoSettings>emptyGen()
        .appendConditionally(
            HashCodeContent::hasArrayProperty,
            Generator.ofWriterFunction(w -> w.ref(JavaRefs.JAVA_UTIL_ARRAYS)));
  }

  /** Content for the generation of the hashCode method. */
  @PojoBuilder(builderName = "HashCodeContentBuilder")
  @Value
  public static class HashCodeContent {
    PList<JavaPojoMember> members;
    Optional<JavaAdditionalProperties> additionalProperties;

    private boolean hasArrayProperty() {
      return members.exists(m -> m.getJavaType().isJavaArray());
    }
  }

  @Value
  private static class HashCodeMember {
    JavaPojoMember member;

    private PList<String> toHashCodeMethodArguments() {
      return member.createFieldNames().map(JavaIdentifier::asString).map(this::mapArrayArguments);
    }

    private String mapArrayArguments(String argument) {
      return member.getJavaType().isJavaArray()
              && argument.equals(member.getName().asIdentifier().asString())
          ? String.format("Arrays.hashCode(%s)", argument)
          : argument;
    }
  }
}
