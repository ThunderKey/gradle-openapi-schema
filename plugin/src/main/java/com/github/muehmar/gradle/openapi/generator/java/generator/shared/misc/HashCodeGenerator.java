package com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc;

import static com.github.muehmar.gradle.openapi.util.Functions.allExceptFirst;
import static io.github.muehmar.codegenerator.java.JavaModifier.PUBLIC;

import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.AnnotationGenerator;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.java.JavaGenerators;
import io.github.muehmar.codegenerator.writer.Writer;
import io.github.muehmar.pojobuilder.annotations.PojoBuilder;
import lombok.Value;

public class HashCodeGenerator {
  private HashCodeGenerator() {}

  public static Generator<HashCodeContent, PojoSettings> hashCodeMethod() {
    final Generator<HashCodeContent, PojoSettings> method =
        JavaGenerators.<HashCodeContent, PojoSettings>methodGen()
            .modifiers(PUBLIC)
            .noGenericTypes()
            .returnType("int")
            .methodName("hashCode")
            .noArguments()
            .content(hashCodeMethodContent())
            .build();
    return AnnotationGenerator.<HashCodeContent, PojoSettings>override()
        .append(method)
        .append(arraysRefGenerator());
  }

  private static Generator<HashCodeContent, PojoSettings> hashCodeMethodContent() {
    return (content, s, w) -> {
      final Writer writerStartPrinted = w.println("return Objects.hash(");

      return content
          .getTechnicalPojoMembers()
          .map(HashCodeGenerator::mapToHashCodeArgument)
          .reverse()
          .zipWithIndex()
          .map(allExceptFirst(arg -> arg.concat(",")))
          .reverse()
          .foldLeft(writerStartPrinted, (writer, name) -> writer.tab(1).println(name))
          .println(");")
          .ref(JavaRefs.JAVA_UTIL_OBJECTS);
    };
  }

  private static Generator<HashCodeContent, PojoSettings> arraysRefGenerator() {
    return Generator.<HashCodeContent, PojoSettings>emptyGen()
        .appendConditionally(
            HashCodeContent::hasArrayProperty,
            Generator.ofWriterFunction(w -> w.ref(JavaRefs.JAVA_UTIL_ARRAYS)));
  }

  /** Content for the generation of the hashCode method. */
  @PojoBuilder(builderName = "HashCodeContentBuilder")
  @Value
  public static class HashCodeContent {
    PList<TechnicalPojoMember> technicalPojoMembers;

    private boolean hasArrayProperty() {
      return technicalPojoMembers.exists(m -> m.getJavaType().isJavaArray());
    }
  }

  private static String mapToHashCodeArgument(TechnicalPojoMember member) {
    return member.getJavaType().isJavaArray()
        ? String.format("Arrays.hashCode(%s)", member.getName())
        : member.getName().asString();
  }
}
