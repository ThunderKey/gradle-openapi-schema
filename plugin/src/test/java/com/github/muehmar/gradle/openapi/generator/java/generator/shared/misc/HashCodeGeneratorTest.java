package com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc;

import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.byteArrayMember;
import static com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMembers.requiredDouble;
import static com.github.muehmar.gradle.openapi.generator.settings.TestPojoSettings.defaultTestSettings;
import static io.github.muehmar.codegenerator.writer.Writer.javaWriter;
import static org.junit.jupiter.api.Assertions.assertTrue;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import ch.bluecare.commons.data.PList;
import com.github.muehmar.gradle.openapi.generator.java.JavaRefs;
import com.github.muehmar.gradle.openapi.generator.java.generator.shared.misc.HashCodeGenerator.HashCodeContent;
import com.github.muehmar.gradle.openapi.generator.java.model.JavaPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.TechnicalPojoMember;
import com.github.muehmar.gradle.openapi.generator.java.model.pojo.JavaPojos;
import com.github.muehmar.gradle.openapi.generator.settings.PojoSettings;
import com.github.muehmar.gradle.openapi.snapshot.SnapshotTest;
import io.github.muehmar.codegenerator.Generator;
import io.github.muehmar.codegenerator.writer.Writer;
import org.junit.jupiter.api.Test;

@SnapshotTest
class HashCodeGeneratorTest {
  private Expect expect;

  @Test
  @SnapshotName("allNecessityAndNullabilityVariants")
  void generate_when_allNecessityAndNullabilityVariants_then_correctHashCodeMethod() {
    final Generator<HashCodeContent, PojoSettings> generator = HashCodeGenerator.hashCodeMethod();
    final Writer writer =
        generator.generate(
            JavaPojos.allNecessityAndNullabilityVariants().getHashCodeContent(),
            defaultTestSettings(),
            javaWriter());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_OBJECTS::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("arrayPojo")
  void generate_when_arrayPojo_then_correctHashCodeMethod() {
    final Generator<HashCodeContent, PojoSettings> generator = HashCodeGenerator.hashCodeMethod();
    final Writer writer =
        generator.generate(
            JavaPojos.arrayPojo().getHashCodeContent(), defaultTestSettings(), javaWriter());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_OBJECTS::equals));

    expect.toMatchSnapshot(writer.asString());
  }

  @Test
  @SnapshotName("byteArrayMember")
  void generate_when_byteArrayMember_then_correctHashCodeMethod() {
    final Generator<HashCodeContent, PojoSettings> generator = HashCodeGenerator.hashCodeMethod();

    final PList<TechnicalPojoMember> technicalMembers =
        PList.of(byteArrayMember(), requiredDouble()).flatMap(JavaPojoMember::getTechnicalMembers);
    final HashCodeContent hashCodeContent =
        HashCodeContentBuilder.create().technicalPojoMembers(technicalMembers).build();
    final Writer writer = generator.generate(hashCodeContent, defaultTestSettings(), javaWriter());

    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_ARRAYS::equals));
    assertTrue(writer.getRefs().exists(JavaRefs.JAVA_UTIL_OBJECTS::equals));

    expect.toMatchSnapshot(writer.asString());
  }
}
