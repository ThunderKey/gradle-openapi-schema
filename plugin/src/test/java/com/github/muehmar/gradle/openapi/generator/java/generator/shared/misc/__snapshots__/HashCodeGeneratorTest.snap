allNecessityAndNullabilityVariants=[
@Override
public int hashCode() {
  return Objects.hash(
    requiredStringVal,
    requiredNullableStringVal,
    isRequiredNullableStringValPresent,
    optionalStringVal,
    optionalNullableStringVal,
    isOptionalNullableStringValNull,
    additionalProperties
  );
}
]


arrayPojo=[
@Override
public int hashCode() {
  return Objects.hash(
    value
  );
}
]


byteArrayMember=[
@Override
public int hashCode() {
  return Objects.hash(
    Arrays.hashCode(data),
    doubleVal
  );
}
]